File: 131-Palindrome Partitioning-分割回文串/131-Palindrome Partitioning-分割回文串.js
/**
 * @param {string} s
 * @return {string[][]}
 */
 let res = []
 let path = []
 var partition = function(s) {
     res = []
     backtrace(s, 0)
     return res
 };
 
 var backtrace = function(s, index){
     if(index >= s.length){
         res.push(Array.from(path))
         return
     }
 
     for(let i = index; i < s.length; i++){
         let str = s.substring(index, i + 1)
         if(isPalindrome(str)){
             path.push(str)
             backtrace(s, i + 1)
             path.pop()
         }else{
             continue
         }
 
     }
 }
 
 var isPalindrome = function(arr){
     let len = arr.length
     if(len === 0) return false
     if(len === 1) return true
 
     for(let i = 0; i < len; i++){
         if(arr[i] != arr[len - i - 1]){
             return false
         }
     }
     return true
 }

File: 131-Palindrome Partitioning-分割回文串/131-Palindrome Partitioning-分割回文串.java
class Solution {
    List<List<String>> res;
    LinkedList<String> path = new LinkedList<String>();
    public List<List<String>> partition(String s) {
    	res = new ArrayList<>();
    	backtrace(s, 0);
    	return res;
    }
    
    public void backtrace(String s, int index) {
    	if(index >= s.length()) {
    		res.add(new ArrayList(path));
    		return;
    	}
    	
    	for(int i = index; i < s.length(); i++) {
    		String str = s.substring(index, i + 1);
    		if(isPalindrome(str)) {
    			path.add(str);
    			backtrace(s, i + 1);
    			path.removeLast();
    		}else {
    			continue;
    		}
    	}
    }
    
    public boolean isPalindrome(String s) {
    	int len = s.length();
    	if(len == 0) return false;
    	if(len == 1) return true;
    	for(int i = 0; i < len; i++) {
    		if(s.charAt(i) != s.charAt(len - i - 1)) {
    			return false;
    		}
    	}
    	return true;
    }
}